### ISSUE TITLE
Can TensorRT 'FAIL' to compress the original network? (output value of TensorRT is difference from the original pre-trained network)

### ISSUE THREAD

**Issue by nonstop1962** [NVIDIA/TensorRT#202]
## Description
Hi, I build my own network using tensorRT C++ API from scratch. I modified the network construction part of the sample code 'sampleMNISTAPI' with my own network (constructNetwork() function in the sample code). 

I'm pretty sure that my implementation of network construction using C++ API has no error, but I found that tensorRT sometimes fails to reproduce the output value of the original pre-trained network when the same input data is given.

Below is what I've done to verify this issue. 
1. First, I using tensorRT C++ API for constructing the network architecture from scratch (Squeezenet). There are no external plugins in this network architecture (I used only `IConvolutionLayer`, `IActivationLayer`, `IPoolingLayer`, `IConcatenationLayer`, and `ISoftMaxLayer`). 
```

bool SampleMNISTAPI::constructSqueezeNetNetworkDebug(SampleUniquePtr<nvinfer1::IBuilder>& builder, SampleUniquePtr<nvinfer1::INetworkDefinition>& network, SampleUniquePtr<nvinfer1::IBuilderConfig>& config, int layerIdx)
{
	//////////////////////////// First layers ////////////////////////////
	// Create input tensor of shape { 1, inputChannel, inputH, inputW}
	ITensor* data = network->addInput(
		mParams.inputTensorNames[0].c_str(), DataType::kFLOAT, Dims3{ mParams.inputChannel, mParams.inputH, mParams.inputW });
	assert(data);
	data->setName("data");
	// Add convolution layer with 64 outputs and a 3x3 filter with 2x2 stride.
	IConvolutionLayer* conv1 = network->addConvolution(*data, 64, DimsHW{ 3, 3 }, mWeightMap["conv0"], mWeightMap["bias0"]);
	assert(conv1);
	conv1->setStride(DimsHW{ 2, 2 });
	conv1->setName("conv1");
	// Add activation layer using the ReLU algorithm.
	IActivationLayer* relu_conv1 = network->addActivation(*conv1->getOutput(0), ActivationType::kRELU);
	assert(relu_conv1);
	relu_conv1->setName("relu_conv1");
	// Add max pooling layer with stride of 2x2 and kernel size of 3x3>
	IPoolingLayer* pool1 = network->addPooling(*relu_conv1->getOutput(0), PoolingType::kMAX, DimsHW{ 3, 3 });
	assert(pool1);
	pool1->setStride(DimsHW{ 2, 2 });
	pool1->setName("pool1");


	//////////////////////////// Fire 2 ////////////////////////////
	// Add convolution layer with 16 outputs and a 1x1 filter with 1x1 stride.
	IConvolutionLayer* fire2_squeeze1x1 = network->addConvolution(*pool1->getOutput(0), 16, DimsHW{ 1, 1 }, mWeightMap["conv1"], mWeightMap["bias1"]);
	assert(fire2_squeeze1x1);
	fire2_squeeze1x1->setStride(DimsHW{ 1, 1 });
	fire2_squeeze1x1->setName("fire2_squeeze1x1");

... (Do similar things for remaining network)

	// Add softmax layer to determine the probability.
	ISoftMaxLayer* prob = network->addSoftMax(*global_average_pooling2d_1->getOutput(0));
	assert(prob);
	prob->setName("Softmax");

	ILayer* targetOutputLayer = network->getLayer(layerIdx);
	targetOutputLayer->getOutput(0)->setName(mParams.outputTensorNames[0].c_str());
	network->markOutput(*targetOutputLayer->getOutput(0));

	// Build engine
	builder->setMaxBatchSize(mParams.batchSize);
	//config->setMaxWorkspaceSize(16_MiB);
	config->setMaxWorkspaceSize(8_GiB);

	samplesCommon::enableDLA(builder.get(), config.get(), mParams.dlaCore);
	mEngine = std::shared_ptr<nvinfer1::ICudaEngine>(builder->buildEngineWithConfig(*network, *config), samplesCommon::InferDeleter());
```
As you can see in the last part of the above code, I marked the output layer using `layerIdx` parameter (`layerIdx` parameter is varied from `0` to `(number of total layers-1)`). By doing this, I believe that I can construct various structures of network (e.g. `layerIdx = 0` --> network with only one convolutional layer, `layerIdx = (number of total layers-1)` --> full squeezeNet).

2. Using the compressed model that is stored in the `mEngine` in the above code, I run the inference for the same input data and save the output value (slightly modify the `infer()` function in the sampleMNISTAPI code). 

3. Finally, I compare the output of the original pre-trained network and the compressed model from the tensorRT for every single layer (difference of the output values is computed using np.linalg.norm(original_output- tensorRT_output)). The results are shown as follows:
```
================================
Layer name:  conv1
Network output difference:  10530.271
================================

================================
Layer name:  relu_conv1
Network output difference:  10470.494
================================

================================
Layer name:  pool1
Network output difference:  7086.1274
================================

================================
Layer name:  fire2_squeeze1x1
Network output difference:  122.3007
================================

================================
Layer name:  fire2_relu_squeeze1x1
Network output difference:  0.0014203875
================================

================================
Layer name:  fire2_expand1x1
Network output difference:  10364.62
================================

================================
Layer name:  fire2_relu_expand1x1
Network output difference:  10299.981
================================

================================
Layer name:  fire2_expand3x3
Network output difference:  1663.5532
================================

================================
Layer name:  fire2_relu_expand3x3
Network output difference:  0.002477309
================================

================================
Layer name:  fire2_concat
Network output difference:  0.0034513343
================================

================================
Layer name:  fire3_squeeze1x1
Network output difference:  99.82256
================================

================================
Layer name:  fire3_relu_squeeze1x1
Network output difference:  0.0037087006
================================

================================
Layer name:  fire3_expand1x1
Network output difference:  6869.336
================================

================================
Layer name:  fire3_relu_expand1x1
Network output difference:  6678.0264
================================

================================
Layer name:  fire3_expand3x3
Network output difference:  2756.992
================================

================================
Layer name:  fire3_relu_expand3x3
Network output difference:  0.0075148246
================================

================================
Layer name:  fire3_concat
Network output difference:  0.009390493
================================

================================
Layer name:  pool3
Network output difference:  10029.8955
================================

================================
Layer name:  fire4_squeeze1x1
Network output difference:  496.88153
================================

================================
Layer name:  fire4_relu_squeeze1x1
Network output difference:  0.007861499
================================

================================
Layer name:  fire4_expand1x1
Network output difference:  6511.7163
================================

================================
Layer name:  fire4_relu_expand1x1
Network output difference:  6157.9697
================================

================================
Layer name:  fire4_expand3x3
Network output difference:  5991.5405
================================

================================
Layer name:  fire4_relu_expand3x3
Network output difference:  0.017509103
================================

================================
Layer name:  fire4_concat
Network output difference:  0.020673396
================================

================================
Layer name:  fire5_squeeze1x1
Network output difference:  1329.483
================================

================================
Layer name:  fire5_relu_squeeze1x1
Network output difference:  0.019874217
================================

================================
Layer name:  fire5_expand1x1
Network output difference:  11474.466
================================

================================
Layer name:  fire5_relu_expand1x1
Network output difference:  9779.374
================================

================================
Layer name:  fire5_expand3x3
Network output difference:  14287.976
================================

================================
Layer name:  fire5_relu_expand3x3
Network output difference:  0.029271929
================================

================================
Layer name:  fire5_concat
Network output difference:  0.039569486
================================

================================
Layer name:  pool5
Network output difference:  8051.586
================================

================================
Layer name:  fire6_squeeze1x1
Network output difference:  2301.3528
================================

================================
Layer name:  fire6_relu_squeeze1x1
Network output difference:  0.023320064
================================

================================
Layer name:  fire6_expand1x1
Network output difference:  11248.362
================================

================================
Layer name:  fire6_relu_expand1x1
Network output difference:  9016.117
================================

================================
Layer name:  fire6_expand3x3
Network output difference:  16513.094
================================

================================
Layer name:  fire6_relu_expand3x3
Network output difference:  0.04112033
================================

================================
Layer name:  fire6_concat
Network output difference:  0.04808631
================================

================================
Layer name:  fire7_squeeze1x1
Network output difference:  2704.302
================================

================================
Layer name:  fire7_relu_squeeze1x1
Network output difference:  0.04152312
================================

================================
Layer name:  fire7_expand1x1
Network output difference:  13887.17
================================

================================
Layer name:  fire7_relu_expand1x1
Network output difference:  9521.085
================================

================================
Layer name:  fire7_expand3x3
Network output difference:  27407.674
================================

================================
Layer name:  fire7_relu_expand3x3
Network output difference:  0.071296364
================================

================================
Layer name:  fire7_concat
Network output difference:  0.083069086
================================

================================
Layer name:  fire8_squeeze1x1
Network output difference:  6611.349
================================

================================
Layer name:  fire8_relu_squeeze1x1
Network output difference:  0.063248456
================================

================================
Layer name:  fire8_expand1x1
Network output difference:  15210.469
================================

================================
Layer name:  fire8_relu_expand1x1
Network output difference:  2669.806
================================

================================
Layer name:  fire8_expand3x3
Network output difference:  49885.887
================================

================================
Layer name:  fire8_relu_expand3x3
Network output difference:  0.10497102
================================

================================
Layer name:  fire8_concat
Network output difference:  0.11630318
================================

================================
Layer name:  fire9_squeeze1x1
Network output difference:  10452.268
================================

================================
Layer name:  fire9_relu_squeeze1x1
Network output difference:  0.09190958
================================

================================
Layer name:  fire9_expand1x1
Network output difference:  53902.543
================================

================================
Layer name:  fire9_relu_expand1x1
Network output difference:  44.051155
================================

================================
Layer name:  fire9_expand3x3
Network output difference:  98443.22
================================

================================
Layer name:  fire9_relu_expand3x3
Network output difference:  0.10746483
================================

================================
Layer name:  fire9_concat
Network output difference:  0.12190292
================================

================================
Layer name:  last_conv0
Network output difference:  1546.0013
================================

================================
Layer name:  relu_conv10_new
Network output difference:  1.000152
================================

================================
Layer name:  global_average_pooling2d_1
Network output difference:  0.00024093324
================================

================================
Layer name:  Softmax
Network output difference:  5.4285465e-11
================================
```
As you can see, the difference of the final layer (Softmax) is almost zero. I think this means I've done the right way (network construction, weight substitution, same input data...) . 
But differences of other layers are significantly high. 

I want to ask if these results are natural. If these results are natural, is it means that tensorRT has possibility to fail to compress the model for some specific architecture (e.g. network consists only convolutional layer without softmax layer)?

## Environment

**TensorRT Version**: TensorRT-6.0.1.5
**GPU Type**: NVIDIA TITAN Xp
**Nvidia Driver Version**:  425.31
**CUDA Version**: 10.0
**CUDNN Version**: 7.6.3
**Python Version (if applicable)**: 
**TensorFlow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Operating System + Version**: Windows 10

**Comment by rmccorm4** [comment]
I think you're right in thinking that comparing the output is what matters more. I don't know the actual answer, but I know that TensorRT does various optimizations such as layer fusion when building the engine:  https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#layer-fusion

Because of that, I would think your TensorRT engine won't really have a 1-1 comparison that you can do for every layer, because the intermediate structure may be different between the engine and the original model.

**Comment by nonstop1962** [comment]
Thanks for your answer. 
Then, can I understand that tensorRT DOES NOT always guarantee the same output value as the original network? Or are there any tensorRT methods to guarantee the same result as the original network?

I think it is important to know the above things to apply to the real applications where performance is important.

**Comment by pranavm-nvidia** [comment]
TensorRT does guarantee that the output value will match the original network. 

How are you computing `Network output difference`? If it's a summation of the absolute difference, you may just be accumulating fp rounding errors, which would explain why you see larger differences for the layers with larger outputs.

**Comment by nonstop1962** [comment]
@rmccorm4 @pranavm-nvidia Thanks for the reply. Actually, there were some problems in my original network and I saw the wrong values as a reference of comparison. After fixing the problem in my part, the outputs are the same. Since it is solved, I will close this issue.
